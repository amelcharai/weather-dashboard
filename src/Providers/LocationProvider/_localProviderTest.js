import { useLocation } from "."
export const LocalProviderTest = () => {
  const {
    suggestedLocations,
    inputValue,
    searchCity,
    searchGeoCode,
    setSuggestedLocations,
    setInputValue,
    onChangeInputValue,
    submitLocation,
    resetSearch,
    setSearchCity,
    setSearchGeoCode
  } = useLocation();

  const mockEvnet = {
      target: {
        value: 'mock change value'
      }
  }
  
  return (
  
      <div>
        <div data-testid='suggestedLocations' onClick={() => setSuggestedLocations('mock city')}>{suggestedLocations}</div>
        <div data-testid='inputValue' onClick={() => setInputValue('mock input')}>{inputValue}</div>
        <div data-testid='searchCity' onClick={() => setSearchCity('mock search city')}>{searchCity}</div>
        <div data-testid='searchGeoCode' onClick={() => setSearchGeoCode('mock geo code')}>{searchGeoCode}</div>
        <div data-testid='onChangeInputValue' onClick={() => onChangeInputValue(mockEvnet)}>{inputValue}</div>
      <div data-testid='resetSearch' onClick={() => resetSearch()}>{`${searchCity},${searchGeoCode}`}</div>
      <div data-testid='submitLocation' onClick={() => submitLocation()}>{searchCity}</div>

      </div>
  )
}