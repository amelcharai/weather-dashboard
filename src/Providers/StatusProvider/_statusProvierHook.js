/* istanbul ignore file */
/**
 * This file should be inside of __tests__ folder,
 * however, test configure this time doesn't skip this file. 
 * so temporary purpose, I just put it out side of test folder.
 */
import { useStatus } from "."

export const StatusProviderTest = () => {
  const {
    error,
    updateWeather,
    isLoading,
    isValidAccess,
    setError,
    setWeatherRequest,
    setLoading,
    setValidAccess
  } = useStatus();
  return (
    <div>
      <div data-testid='error' onClick={() => setError('error')}>{`${error}`}</div>
      <div data-testid='weatherRequest' onClick={() => setWeatherRequest(true)}>{`${updateWeather}`}</div>
      <div data-testid='loading' onClick={() => setLoading(true)}>{`${isLoading}`}</div>
      <div data-testid='validAccess' onClick={() => setValidAccess(true)}>{`${isValidAccess}`}</div>
    </div>
  )
}