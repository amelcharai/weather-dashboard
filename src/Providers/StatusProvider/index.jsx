import { createContext, useContext, useState, useEffect } from 'react';

export const StatusContext = createContext(null);

export const StatusProvider = ({ children }) => {
  const [error, setError] = useState(null);
  const [updateWeather, setWeatherRequest] = useState(false);
  const [isLoading, setLoading] = useState(false);
  const [isValidAccess, setValidAccess] = useState(false);
  useEffect(() => {
    if (error) {
      setLoading(false);
    }
  }, [error]);

  const statusContextValue = {
    error,
    updateWeather,
    isLoading,
    isValidAccess,
    setError,
    setWeatherRequest,
    setLoading,
    setValidAccess
  }
  return (
    <StatusContext.Provider value={statusContextValue}>
      {children}
    </StatusContext.Provider>
  )
}

export const useStatus = () => {
  return useContext(StatusContext);
}