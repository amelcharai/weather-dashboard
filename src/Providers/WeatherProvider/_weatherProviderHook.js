/* istanbul ignore file */
/**
 * This file should be inside of __tests__ folder,
 * however, test configure this time doesn't skip this file. 
 * so temporary purpose, I just put it out side of test folder.
 */
import { useWeather } from ".";
import { useStatus } from "../StatusProvider";
import { useLocation } from "../LocationProvider";
import { useEffect } from "react";

export const WeatherProviderTest = () => {
  const {
    current,
    location,
    forecast,
    isMetric,
    setIsMetric } = useWeather();
  const { setWeatherRequest, setError } = useStatus();
  const { setSearchCity } = useLocation();
  useEffect(() => {
    setError(null);
    setSearchCity('Calgary');
  },[setSearchCity])
  
  return (
    <div>
      <div data-testId='metric' onClick={() => setIsMetric(false)}>{`${isMetric}`}</div>
      <div data-testId='weatherRequest' onClick={() => setWeatherRequest(true)}>weather request</div>
      {current && (<div data-testId='current'>{Object.keys(current).length}</div>)}
      {location && (<div data-testId='location'>{Object.keys(location).length}</div>)}
      {forecast && (<div data-testId='forecast'>{Object.keys(forecast).length}</div>)}
    </div>
  )
}