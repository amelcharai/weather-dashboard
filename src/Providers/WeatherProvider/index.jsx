import { createContext, useContext, useReducer, useEffect, useState } from 'react';
import axios from 'axios';
import { weatherReducer } from '../../reducers/weatherReducer';
import { useStatus } from '../StatusProvider';
import { useLocation } from '../LocationProvider';

const WeatherContext = createContext(null);
const WeatherDispatchContext = createContext(null);

export const WeatherProvider = ({ children }) => {
  const [weather, dispatch] = useReducer(
    weatherReducer,
    {
      status: 'uninit'
    }
  )
  const [isMetric, setIsMetric] = useState(true); // metric or imperial
  const { error, updateWeather, setWeatherRequest, setError, setLoading } = useStatus();
  const { searchCity, searchGeoCode, resetSearch } = useLocation();
  useEffect(() => {
    (async () => {
      try {
        if (!error
          && (searchCity || searchGeoCode)
          && updateWeather
          && process.env.REACT_APP_WEATHER_API_KEY
        ) {
          let searchKey = '';
          if (searchGeoCode) {
            searchKey = `${searchGeoCode.lat},${searchGeoCode.lon}`;
          }
          else if (searchCity) {
            searchKey = searchCity;
          }
          setLoading(true);
          const url = `${process.env.REACT_APP_WEATHER_API_URL}/forecast.json?key=${process.env.REACT_APP_WEATHER_API_KEY}&q=${searchKey}&days=6`;
          const response = await axios({ method: 'GET', url })
          
          const { current, location, forecast } = response.data;
          dispatch({
            type: 'success',
            current,
            location,
            forecast: forecast.forecastday.slice(1),
          });
          setLoading(false);
          resetSearch();
          setWeatherRequest(false);
        }
      } catch (err) {
        const { code, message } = err.response?.data?.error || {};
        setError({
          ...code && message ? { code, message } : err
        });
      };
    })();
  }, [updateWeather, error])
  
  const avelableContextValue = {
    ...weather,
    isMetric,
    setIsMetric
  }

  return (
    <WeatherContext.Provider value={avelableContextValue}>
      <WeatherDispatchContext.Provider value={dispatch}>
        {children}
      </WeatherDispatchContext.Provider>
    </WeatherContext.Provider>
  );
}

export const useWeather = () => {
  return useContext(WeatherContext);
}

export const useWeatherDispatch = () => {
  return useContext(WeatherDispatchContext);
}