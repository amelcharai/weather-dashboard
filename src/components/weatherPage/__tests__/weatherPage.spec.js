import { render, screen } from "@testing-library/react";
import { useGeolocated } from "react-geolocated";
import { BrowserRouter } from "react-router-dom";
import { WeatherProvider, useWeather } from "../../../Providers/WeatherProvider";
import { LocationProvider } from "../../../Providers/LocationProvider";
import { StatusProvider, useStatus } from "../../../Providers/StatusProvider";
import { Weather } from "..";
// eslint-disable-next-line jest/no-mocks-import
import { mockCurrent, mockForecast,mockLocation } from "../../../__mocks__/mockWeather";


jest.mock('../../../Providers/WeatherProvider', () => ({
  ...jest.requireActual('../../../Providers/WeatherProvider'),
  useWeather: () => {
    return {
      current: mockCurrent,
      location: mockLocation,
      forecast: mockForecast,
      isMetric: true,
      status:'success',
      setIsMetric: ()=> jest.fn()
    }
  }
}));

jest.mock('../../../Providers/StatusProvider', () => ({
  ...jest.requireActual('../../../Providers/StatusProvider'),
  useStatus: () => {
    return {
      error: undefined,
      updateWeather: false,
      isLoading: false,
      isValidAccess: true,
      setError: ()=> jest.fn(),
      setWeatherRequest:()=> jest.fn(),
      setLoading:()=> jest.fn(),
      setValidAccess:()=> jest.fn()
    }
  }
}));

jest.mock('react-geolocated', () => ({
  useGeolocated: () => {
    return {
      coords: {
        lat: 0,
        lon: 0
      },
      isGeolocationAvailable: true,
      isGeolocationEnabled: true
    }
  }
}));

describe('Weather page componet', () => {
  const mount = (children) => {
    return render(
      <StatusProvider>
        <LocationProvider>
          <WeatherProvider>
            <BrowserRouter>
              {children}
            </BrowserRouter>
          </WeatherProvider>
        </LocationProvider>
      </StatusProvider>
    )
  }
  
  it('Should display as expected', () => {
    const { container } = mount(
      <Weather />
    )
   
    screen.getByText(/Humidity/);
    screen.getByText(/Wind Speed/);
    screen.getByText(/Testerazo/);
    screen.getByText(/Wind Speed/);
    expect(container).toMatchSnapshot();

  })
})
