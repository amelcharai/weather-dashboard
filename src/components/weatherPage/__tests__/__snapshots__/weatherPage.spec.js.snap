// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather page componet Should display as expected 1`] = `
<div>
  <div
    class="container"
    id="weatherPage"
  >
    <div
      id="header"
    >
      <div
        class="cityName"
      >
        Testerazo (Puente Cajon De Gatos) San Luis PotosÃ­, Mexico
      </div>
      <div
        class="btn-group"
        role="group"
      >
        <input
          autocomplete="off"
          checked=""
          class="btn-check"
          id="metric"
          name="radio"
          type="radio"
          value=""
        />
        <label
          class="btn btn-dark"
          for="metric"
          tabindex="0"
        >
          Metric
        </label>
        <input
          autocomplete="off"
          class="btn-check"
          id="imperial"
          name="radio"
          type="radio"
          value=""
        />
        <label
          class="btn btn-outline-dark"
          for="imperial"
          tabindex="0"
        >
          Imperial
        </label>
      </div>
      <div>
        Wed, Apr 17 2024
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-sm-3"
      >
        <div
          class="weatherIcon"
        >
          <img
            alt="Sunny"
            src="/weather-icons/1000.svg"
          />
          <div
            class="text shiftUpText"
          >
            Sunny
          </div>
        </div>
      </div>
      <div
        class="col-sm-9"
      >
        <div
          class="currentWeatherInfoContainer"
        >
          <div
            class="temperature"
          >
            <span
              class="temp"
            >
              34.3
            </span>
            <span
              class="degree"
            >
              °C
            </span>
          </div>
          <div
            class="dividedInfo"
          >
            <div
              class="top"
            >
              <div
                class="label"
              >
                Humidity
                 :
              </div>
              <div
                class="content"
              >
                34 %
              </div>
            </div>
            <div
              class="bottom"
            >
              <div
                class="label"
              >
                Wind Speed
                 :
              </div>
              <div
                class="content"
              >
                17.6kph E
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col"
      >
        <section
          id="weekForcast"
        >
          <div
            class="forcastCard"
          >
            <div>
              Thu 18
            </div>
            <div>
              <div
                class="weatherIcon"
              >
                <img
                  alt="Patchy rain nearby"
                  src="/weather-icons/1063.svg"
                />
                <div
                  class="text  "
                >
                  Patchy rain nearby
                </div>
              </div>
            </div>
            <div
              class="tempContainer"
            >
              <div
                class="dividedInfo"
              >
                <div
                  class="top"
                >
                  <div
                    class="label"
                  >
                    Hi
                     :
                  </div>
                  <div
                    class="content"
                  >
                    39.3
                  </div>
                </div>
                <div
                  class="bottom"
                >
                  <div
                    class="label"
                  >
                    Lo
                     :
                  </div>
                  <div
                    class="content"
                  >
                    22.8
                  </div>
                </div>
              </div>
              <div
                class="degree"
              >
                °C
              </div>
            </div>
          </div>
          <div
            class="forcastCard"
          >
            <div>
              Fri 19
            </div>
            <div>
              <div
                class="weatherIcon"
              >
                <img
                  alt="Sunny"
                  src="/weather-icons/1000.svg"
                />
                <div
                  class="text  "
                >
                  Sunny
                </div>
              </div>
            </div>
            <div
              class="tempContainer"
            >
              <div
                class="dividedInfo"
              >
                <div
                  class="top"
                >
                  <div
                    class="label"
                  >
                    Hi
                     :
                  </div>
                  <div
                    class="content"
                  >
                    39.3
                  </div>
                </div>
                <div
                  class="bottom"
                >
                  <div
                    class="label"
                  >
                    Lo
                     :
                  </div>
                  <div
                    class="content"
                  >
                    21
                  </div>
                </div>
              </div>
              <div
                class="degree"
              >
                °C
              </div>
            </div>
          </div>
          <div
            class="forcastCard"
          >
            <div>
              Sat 20
            </div>
            <div>
              <div
                class="weatherIcon"
              >
                <img
                  alt="Moderate rain"
                  src="/weather-icons/1189.svg"
                />
                <div
                  class="text  "
                >
                  Moderate rain
                </div>
              </div>
            </div>
            <div
              class="tempContainer"
            >
              <div
                class="dividedInfo"
              >
                <div
                  class="top"
                >
                  <div
                    class="label"
                  >
                    Hi
                     :
                  </div>
                  <div
                    class="content"
                  >
                    30.9
                  </div>
                </div>
                <div
                  class="bottom"
                >
                  <div
                    class="label"
                  >
                    Lo
                     :
                  </div>
                  <div
                    class="content"
                  >
                    22
                  </div>
                </div>
              </div>
              <div
                class="degree"
              >
                °C
              </div>
            </div>
          </div>
          <div
            class="forcastCard"
          >
            <div>
              Sun 21
            </div>
            <div>
              <div
                class="weatherIcon"
              >
                <img
                  alt="Patchy rain nearby"
                  src="/weather-icons/1063.svg"
                />
                <div
                  class="text  "
                >
                  Patchy rain nearby
                </div>
              </div>
            </div>
            <div
              class="tempContainer"
            >
              <div
                class="dividedInfo"
              >
                <div
                  class="top"
                >
                  <div
                    class="label"
                  >
                    Hi
                     :
                  </div>
                  <div
                    class="content"
                  >
                    22
                  </div>
                </div>
                <div
                  class="bottom"
                >
                  <div
                    class="label"
                  >
                    Lo
                     :
                  </div>
                  <div
                    class="content"
                  >
                    20.6
                  </div>
                </div>
              </div>
              <div
                class="degree"
              >
                °C
              </div>
            </div>
          </div>
          <div
            class="forcastCard"
          >
            <div>
              Mon 22
            </div>
            <div>
              <div
                class="weatherIcon"
              >
                <img
                  alt="Partly Cloudy "
                  src="/weather-icons/1003.svg"
                />
                <div
                  class="text  "
                >
                  Partly Cloudy 
                </div>
              </div>
            </div>
            <div
              class="tempContainer"
            >
              <div
                class="dividedInfo"
              >
                <div
                  class="top"
                >
                  <div
                    class="label"
                  >
                    Hi
                     :
                  </div>
                  <div
                    class="content"
                  >
                    34.1
                  </div>
                </div>
                <div
                  class="bottom"
                >
                  <div
                    class="label"
                  >
                    Lo
                     :
                  </div>
                  <div
                    class="content"
                  >
                    21.3
                  </div>
                </div>
              </div>
              <div
                class="degree"
              >
                °C
              </div>
            </div>
          </div>
        </section>
      </div>
    </div>
  </div>
</div>
`;
